src/types/reducers.js

```diff
@@ -41,21 +43,63 @@ export type ProfileViewState = {
     tabOrder: number[],
     rightClickedThread: ThreadIndex,
   },
-  profile: Profile,
+  profile: Profile | null,
 };
 
 export type AppViewState =
-  | {| phase: string |}
-  | {
-      phase: 'INITIALIZING',
-      additionalData: { attempt: Attempt | null, message: string },
-    }
-  | { phase: 'FATAL_ERROR', error: Error };
+  | {| +phase: 'ROUTE_NOT_FOUND' |}
+  | {| +phase: 'DATA_LOADED' |}
+  | {| +phase: 'FATAL_ERROR', +error: Error |}
+  | {|
+      +phase: 'INITIALIZING',
+      +additionalData?: {| +attempt: Attempt | null, +message: string |},
+    |};
+
+/**
+ * This represents the finite state machine for loading zip files. The phase represents
+ * where the state is now, and next is the phase it can transition to.
+ */
+export type ZipFileState =
+  | {|
+      +phase: 'NO_ZIP_FILE',
+      +zip: null,
+      +zipFilePath: null,
+    |}
+  | {|
+      +phase: 'LIST_FILES_IN_ZIP_FILE',
+      +zip: JSZip,
+      +zipFilePath: null,
+    |}
+  | {|
+      +phase: 'PROCESS_PROFILE_FROM_ZIP_FILE',
+      +zip: JSZip,
+      +zipFilePath: string,
+    |}
+  | {|
+      +phase: 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE',
+      +zip: JSZip,
+      +zipFilePath: string,
+    |}
+  | {|
+      +phase: 'FILE_NOT_FOUND_IN_ZIP_FILE',
+      +zip: JSZip,
+      +zipFilePath: string,
+    |}
+  | {|
+      +phase: 'VIEW_PROFILE_IN_ZIP_FILE',
+      +zip: JSZip,
+      +zipFilePath: string,
```

Comment:
I agree that `zipFilePath` is awkward. I'd prefer to keep the longer name, so that the selectors and various usages can have an unambiguous name. I renamed it to `pathInZipFile` to be a bit more unambiguous.